File: app/src/main/java/com/hsilighting/pagify/core/PageMapping.kt
package com.hsilighting.pagify.core

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect

/**
 * Where a point on the page lands on screen, and back again.
 *
 * Marks are stored in page points, which is what lets one stroke survive a zoom,
 * a rotation and a reopen. Everything that draws or hit-tests one has to convert,
 * and until the reader could rotate that conversion was a multiply and an add —
 * so it was written inline, twice per call site, in about twenty places.
 *
 * Rotation is what makes it worth a type. A turned page is not a scaled page: the
 * axes swap and one of them flips, and getting the flip backwards puts every mark
 * on the wrong side of the sheet in a way that still looks like a page of marks.
 * Pure, and here rather than in the layer that draws, because that is the only way
 * to test it without a device.
 *
 * @param scale pixels per point, against the page *as laid out* — so against its
 *   turned width, which is what the box on screen is measured in.
 * @param origin top-left of the drawn page, in the space touches arrive in.
 * @param quarterTurns view rotation, clockwise, matching the renderer's.
 * @param pageWidthPoints the page's own width, before any turn.
 * @param pageHeightPoints likewise its own height.
 */
data class PageMapping(
    val scale: Float,
    val origin: Offset = Offset.Zero,
    val quarterTurns: Int = 0,
    val pageWidthPoints: Float = 0f,
    val pageHeightPoints: Float = 0f,
) {

    /** Whether anything can be placed at all; a page not yet measured cannot. */
    val isUsable: Boolean get() = scale > 0f

    private val turns: Int get() = ((quarterTurns % 4) + 4) % 4

    /** A page point, in the pixels of the page as drawn. */
    fun toScreen(point: Offset): Offset = turnPoint(point) * scale + origin

    /** A touch, in the page's own points. */
    fun toPage(position: Offset): Offset {
        if (!isUsable) return Offset.Zero
        return unturnPoint((position - origin) / scale)
    }

    /**
     * A rectangle of page points, as the box it covers on screen.
     *
     * Both corners go through the turn and the result is normalised, because a
     * quarter turn sends a top-left corner to a top-right one — keeping the
     * corners in their original roles would give a rectangle of negative width,
     * which draws as nothing at all.
     */
    fun toScreen(rect: Rect): Rect {
        val a = toScreen(Offset(rect.left, rect.top))
        val b = toScreen(Offset(rect.right, rect.bottom))
        return Rect(
            left = minOf(a.x, b.x),
            top = minOf(a.y, b.y),
            right = maxOf(a.x, b.x),
            bottom = maxOf(a.y, b.y),
        )
    }

    /** A length in points, in pixels. Rotation does not change a length. */
    fun toScreen(lengthPoints: Float): Float = lengthPoints * scale

    /** A length in pixels, in page points. */
    fun toPage(lengthPixels: Float): Float = if (isUsable) lengthPixels / scale else 0f

    /**
     * A page point in the turned page's own point space.
     *
     * Clockwise, to match `PdfPageRenderRotation` in the engine: after a quarter
     * turn the page's top-left corner is at the top *right*, so x comes from the
     * height and y comes from x.
     */
    private fun turnPoint(point: Offset): Offset = when (turns) {
        1 -> Offset(pageHeightPoints - point.y, point.x)
        2 -> Offset(pageWidthPoints - point.x, pageHeightPoints - point.y)
        3 -> Offset(point.y, pageWidthPoints - point.x)
        else -> point
    }

    /** The inverse of [turnPoint]. */
    private fun unturnPoint(point: Offset): Offset = when (turns) {
        1 -> Offset(point.y, pageHeightPoints - point.x)
        2 -> Offset(pageWidthPoints - point.x, pageHeightPoints - point.y)
        3 -> Offset(pageWidthPoints - point.y, point.x)
        else -> point
    }

    companion object {
        /** Nothing measured yet: every conversion is a no-op the caller can skip. */
        val Unmeasured = PageMapping(scale = 0f)
    }
}

File: ios/Sources/PageMapping.swift
import CoreGraphics
import Foundation

/// Between a page's own points and the pixels it is drawn at. Android's
/// `core/PageMapping.kt`.
///
/// One type for both places a page is drawn. The list draws it at the top-left of
/// its row; the magnified view draws it wherever the reader has panned it to. The
/// only difference is the origin — and it is exactly that difference that made
/// the tools stop working when zoomed, because the drawing layer there assumed an
/// origin of zero.
struct PageMapping {
    /// Pixels per page point.
    var scale: CGFloat
    /// Where the page's top-left corner actually sits on screen.
    var origin: CGPoint = .zero
    var quarterTurns: Int = 0
    /// The page's own size, upright — the space marks are stored in.
    var pageWidthPoints: CGFloat = 0
    var pageHeightPoints: CGFloat = 0

    static let unmeasured = PageMapping(scale: 0)

    /// Whether anything can be placed at all; a page not yet measured cannot.
    var isUsable: Bool { scale > 0 }

    private var turns: Int { ((quarterTurns % 4) + 4) % 4 }

    /// A page point, in the pixels of the page as drawn.
    func toScreen(_ point: CGPoint) -> CGPoint {
        let turned = turn(point)
        return CGPoint(x: turned.x * scale + origin.x, y: turned.y * scale + origin.y)
    }

    /// A touch, in the page's own points.
    func toPage(_ position: CGPoint) -> CGPoint {
        guard isUsable else { return .zero }
        return unturn(CGPoint(x: (position.x - origin.x) / scale,
                              y: (position.y - origin.y) / scale))
    }

    /// A page point, pulled back onto the page if it has left it.
    ///
    /// Held at the edge rather than dropped: a stroke taken past the margin should
    /// run along it, which is what every drawing tool does and what a hand
    /// expects. Dropping the points would break the stroke into pieces, and
    /// discarding the gesture would lose a mark someone meant to make.
    ///
    /// Clamping on the way *in* is what keeps the file honest — clipping the
    /// drawing alone would leave the points in the annotation, off the page, where
    /// another viewer might well show them.
    func clampToPage(_ point: CGPoint) -> CGPoint {
        guard pageWidthPoints > 0, pageHeightPoints > 0 else { return point }
        return CGPoint(x: min(max(point.x, 0), pageWidthPoints),
                       y: min(max(point.y, 0), pageHeightPoints))
    }

    private func turn(_ point: CGPoint) -> CGPoint {
        switch turns {
        case 1: return CGPoint(x: pageHeightPoints - point.y, y: point.x)
        case 2: return CGPoint(x: pageWidthPoints - point.x, y: pageHeightPoints - point.y)
        case 3: return CGPoint(x: point.y, y: pageWidthPoints - point.x)
        default: return point
        }
    }

    private func unturn(_ point: CGPoint) -> CGPoint {
        switch turns {
        case 1: return CGPoint(x: point.y, y: pageHeightPoints - point.x)
        case 2: return CGPoint(x: pageWidthPoints - point.x, y: pageHeightPoints - point.y)
        case 3: return CGPoint(x: pageWidthPoints - point.y, y: point.x)
        default: return point
        }
    }
}
